pkg/server/plugin/bundlepublisher/k8sconfigmap/k8sconfigmap.go

```diff
@@ -0,0 +1,271 @@
+			return nil
+		}
+		if cluster.ConfigMapKey == "" {
+			status.ReportErrorf("missing configmap key in cluster configuration")
+			return nil
+		}
+		bundleFormat, err := bundleformat.FromString(cluster.Format)
+		if err != nil {
+			status.ReportErrorf("could not parse bundle format from cluster configuration: %v", err)
+			return nil
+		}
+
+		switch bundleFormat {
+		case bundleformat.JWKS:
+		case bundleformat.SPIFFE:
+		case bundleformat.PEM:
+		default:
+			status.ReportErrorf("bundle format %q is not supported", cluster.Format)
+			return nil
+		}
+		cluster.bundleFormat = bundleFormat
+
+		if cluster.KubeConfigPath == "" {
+			if inClusterUsed {
+				status.ReportError("only one cluster can use in-cluster configuratio (empty kubeconfig_path)")
+				return nil
+			}
+			inClusterUsed = true
+		}
+	}
+
+	return newConfig
+}
+
+// Plugin is the main representation of this bundle publisher plugin.
+type Plugin struct {
+	bundlepublisherv1.UnsafeBundlePublisherServer
+	configv1.UnsafeConfigServer
+
+	config    *Config
+	configMtx sync.RWMutex
+
+	bundle    *types.Bundle
+	bundleMtx sync.RWMutex
+
+	hooks pluginHooks
+	log   hclog.Logger
+}
+
+// SetLogger sets a logger in the plugin.
+func (p *Plugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+// Configure configures the plugin.
+func (p *Plugin) Configure(ctx context.Context, req *configv1.ConfigureRequest) (*configv1.ConfigureResponse, error) {
+	newConfig, _, err := pluginconf.Build(req, buildConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range newConfig.Clusters {
+		k8sClient, err := p.hooks.newK8sClientFunc(newConfig.Clusters[i].KubeConfigPath)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to create Kubernetes client: %v", err)
+		}
+		newConfig.Clusters[i].k8sClient = k8sClient
+	}
+
+	p.setConfig(newConfig)
+	p.setBundle(nil)
+	return &configv1.ConfigureResponse{}, nil
+}
+
+// PublishBundle puts the bundle in the configured Kubernetes ConfigMap.
+func (p *Plugin) PublishBundle(ctx context.Context, req *bundlepublisherv1.PublishBundleRequest) (*bundlepublisherv1.PublishBundleResponse, error) {
+	config, err := p.getConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	if req.Bundle == nil {
+		return nil, status.Error(codes.InvalidArgument, "missing bundle in request")
+	}
+
+	currentBundle := p.getBundle()
+	if proto.Equal(req.Bundle, currentBundle) {
+		// Bundle not changed. No need to publish.
+		return &bundlepublisherv1.PublishBundleResponse{}, nil
+	}
+
+	formatter := bundleformat.NewFormatter(req.Bundle)
+
+	for _, cluster := range config.Clusters {
+		bundleBytes, err := formatter.Format(cluster.bundleFormat)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "could not format bundle: %v", err.Error())
+		}
+
+		cm, err := cluster.k8sClient.GetConfigMap(ctx, cluster.Namespace, cluster.ConfigMapName)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to get ConfigMap: %v", err)
```

Comment:
Two plugin behaviour questions here:
* Should we continue instead of returning? An error with one of the clusters shouldn't affect any other clusters. See https://github.com/spiffe/spire/issues/5872 for a report of this being an issue.
* Should we create the ConfigMap if it doesn't exist? See https://github.com/spiffe/spire/issues/4976 for a request for that.

Reply:
I'd love to get the configmap object out of the chart. k8s really doesnt do well when there are two owners of an object, and having helm be one owner and the publisher be another has caused issues.